drivers: adc: ads1x1x: add support for RDY pin

efra-mx-aqua opened the pull request:
Some hardware configuration require rely on the ALERT/RDY pin to know when and ADC conversion is completed. 

* A pin called `int-gpios` can optionally be define in the DT.
* The polling thread is left as fallback, when the pin is not defined.
* The semaphore is only used if the thread is running 
* It introduces a new DT macro `DT_ANY_COMPAT_HAS_PROP_STATUS_OKAY`
Fixes #57952

efra-mx-aqua committed:
drivers: adc: ads1x1x: add support for RDY pin

Some hardware configuration require rely on the ALERT/RDY
pin to know when and ADC conversion is completed.

The polling thread is left as fallback, when the pin
is not defined.

Signed-off-by: Efrain Calderon <efrain.calderon@aquarobur.com>

efra-mx-aqua commented on drivers/adc/adc_ads1x1x.c:
```diff
@@ -433,8 +466,7 @@ static int ads1x1x_validate_sequence(const struct device *dev, const struct adc_
 
 	if (sequence->oversampling) {
-		LOG_ERR("oversampling not supported");
-		return -ENOTSUP;
```
remove this

efra-mx-aqua replied:
done

efra-mx-aqua commented:
This ready for review

efra-mx-aqua commented on drivers/adc/adc_ads1x1x.c:
```diff
@@ -583,6 +731,7 @@ static const struct adc_driver_api ads1x1x_api = {
 #define ADS1X1X_INIT(t, n, odr_delay_us, res, mux, pgab)                                           \
 	static const struct ads1x1x_config ads##t##_config_##n = {                                 \
 		.bus = I2C_DT_SPEC_GET(DT_INST_ADS1X1X(n, t)),                                     \
+		.alert_rdy = GPIO_DT_SPEC_INST_GET_OR(n, alert_rdy_gpios, {0}),                          \
```
```suggestion
		.alert_rdy = GPIO_DT_SPEC_INST_GET_OR(n, alert_rdy_gpios, {0}),                    \
```

XenuIsWatching commented on drivers/adc/adc_ads1x1x.c:
```diff
@@ -138,11 +141,39 @@ struct ads1x1x_data {
+static inline void setup_rdy_pin(const struct device *dev, bool enable)
+{
+	const struct ads1x1x_config *config = dev->config;
+	unsigned int flags = enable
```
use `gpio_flags_t` typedef

efra-mx-aqua replied:
done

XenuIsWatching commented on drivers/adc/adc_ads1x1x.c:
```diff
@@ -138,11 +141,39 @@ struct ads1x1x_data {
+static inline void setup_rdy_interrupt(const struct device *dev, bool enable)
+{
+	const struct ads1x1x_config *config = dev->config;
+	unsigned int flags = enable
```
use `gpio_flags_t` typedef

efra-mx-aqua replied:
done

XenuIsWatching commented on drivers/adc/adc_ads1x1x.c:
```diff
@@ -138,11 +141,39 @@ struct ads1x1x_data {
 };
 
+#ifdef CONFIG_ADC_ADS1X1X_TRIGGER
+static inline void setup_rdy_pin(const struct device *dev, bool enable)
```
nit: naming of function doesn't seem to be consistent with the others, maybe change name to `ads1x1x_setup_rdy_pin`

efra-mx-aqua replied:
done

XenuIsWatching commented on drivers/adc/adc_ads1x1x.c:
```diff
@@ -193,6 +224,33 @@ static int ads1x1x_start_conversion(const struct device *dev)
+	int rc;
+
+	/* set to 1 to enable conversion ALERT/RDY */
+	ads1x1x_read_reg(dev, ADS1X1X_REG_HI_THRESH, &thresh);
```
check return code here

efra-mx-aqua replied:
done

XenuIsWatching commented on drivers/adc/adc_ads1x1x.c:
```diff
@@ -193,6 +224,33 @@ static int ads1x1x_start_conversion(const struct device *dev)
+	}
+
+	/* set to 0 to enable conversion ALERT/RDY */
+	ads1x1x_read_reg(dev, ADS1X1X_REG_LO_THRESH, &thresh);
```
check return code here

efra-mx-aqua replied:
done

XenuIsWatching commented on drivers/adc/adc_ads1x1x.c:
```diff
@@ -544,6 +613,70 @@ static void ads1x1x_acquisition_thread(const struct device *dev)
+	ads1x1x_adc_perform_read(dev);
+}
+
+static void adc_ads1x1x_conv_ready_cb(const struct device *gpio_dev,
```
nit: naming of function doesn't seem to be consistent with the others, maybe change name to `ads1x1x_conv_ready_cb`

efra-mx-aqua replied:
done

XenuIsWatching commented on drivers/adc/adc_ads1x1x.c:
```diff
@@ -124,6 +126,7 @@ enum {
 
 struct ads1x1x_config {
 	struct i2c_dt_spec bus;
+	struct gpio_dt_spec alert_rdy;
```
Shouldn't some of these need an `#ifdef CONFIG_ADC_ADS1X1X_TRIGGER` otherwise, they go unused

efra-mx-aqua replied:
i have changed it, that is consistent with other drivers

XenuIsWatching reviewed (changes requested):
I'm thinking a test case may need to be added for building with the TRIGGER support within https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/build_all/adc

efra-mx-aqua commented on drivers/adc/adc_ads1x1x.c:
```diff
@@ -138,11 +141,39 @@ struct ads1x1x_data {
+			   config->alert_rdy.dt_flags | flags);
+}
+
+static inline void setup_rdy_interrupt(const struct device *dev, bool enable)
```
done

efra-mx-aqua commented:
> I'm thinking a test case may need to be added for building with the TRIGGER support within https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/build_all/adc

I am not familiar with those tests, do you have any pointers? It seems all the drivers are only built with the default configuration

By default it is built with CONFIG=ADC_ADS1X1X_TRIGGER=y

efra-mx-aqua commented:
@XenuIsWatching I have done the changes you requested

XenuIsWatching commented on drivers/adc/adc_ads1x1x.c:
```diff
@@ -138,11 +143,39 @@ struct ads1x1x_data {
 	struct k_thread thread;
+	k_tid_t	tid;
 	bool differential;
+	struct gpio_callback gpio_cb;
+	struct k_work work;
```
I think both of these would also need a `#ifdef CONFIG_ADC_ADS1X1X_TRIGGER`

efra-mx-aqua replied:
> #ifdef DT_ANY_INST_HAS_PROP_STATUS_OKAY(alert_rdy_gpios)

How would that work? `DT_ANY_INST_HAS_PROP_STATUS_OKAY` relies on DT_DRV_COMPAT, but that is defined at the end.

This driver handles different instances

XenuIsWatching replied:
DT_DRV_COMPAT could be moved to the top, it doesn't have to be at the bottom, but I agree that would overcomplicate this PR, and can be done after this is merged

XenuIsWatching reviewed (changes requested):
I have a comment on `#ifdef CONFIG_ADC_ADS1X1X_TRIGGER`, but I'm beginning to think there is a better way to infer this from the dts without needing a kConfig at all. If the dts has the define for the `alert-rdy-gpios`, then it should be implied that it is to use the interrupt pin rather than threaded output.

Maybe just use the macro `#ifdef DT_ANY_INST_HAS_PROP_STATUS_OKAY(alert_rdy_gpios)` rather than the Kconfig. This should allow it for if at least 1 ads1x1x has the alert-rdy-gpios defined, then it will compile all the code need for the interrupt pin without needing to think about enabling or disabling the kconfig

XenuIsWatching commented on drivers/adc/adc_ads1x1x.c:
```diff
@@ -583,6 +740,7 @@ static const struct adc_driver_api ads1x1x_api = {
 #define ADS1X1X_INIT(t, n, odr_delay_us, res, mux, pgab)                                           \
 	static const struct ads1x1x_config ads##t##_config_##n = {                                 \
 		.bus = I2C_DT_SPEC_GET(DT_INST_ADS1X1X(n, t)),                                     \
+		.alert_rdy = GPIO_DT_SPEC_INST_GET_OR(n, alert_rdy_gpios, {0}),                    \
```
you'd also need `#ifdef CONFIG_ADC_ADS1X1X_TRIGGER` around here else I don't think it would compile

efra-mx-aqua replied:
i will fix that. I would have to create a second INIT macro, since I cannot use a compilation switch inside a macro

efra-mx-aqua replied:
something like this:
```
#ifdef CONFIG_ADC_ADS1X1X_TRIGGER
#define ADS1X1X_CONFIG_INIT(t, n, odr_delay_us, res, mux, pgab)                                    \
	static const struct ads1x1x_config ads##t##_config_##n = {                                 \
		.bus = I2C_DT_SPEC_GET(DT_INST_ADS1X1X(n, t)),                                     \
		.alert_rdy = GPIO_DT_SPEC_INST_GET_OR(n, alert_rdy_gpios, {0}),                    \
		.odr_delay = odr_delay_us,                                                         \
		.resolution = res,                                                                 \
		.multiplexer = mux,                                                                \
		.pga = pgab,                                                                       \
	};
#else
#define ADS1X1X_CONFIG_INIT(t, n, odr_delay_us, res, mux, pgab)                                    \
	static const struct ads1x1x_config ads##t##_config_##n = {                                 \
		.bus = I2C_DT_SPEC_GET(DT_INST_ADS1X1X(n, t)),                                     \
		.odr_delay = odr_delay_us,                                                         \
		.resolution = res,                                                                 \
		.multiplexer = mux,                                                                \
		.pga = pgab,                                                                       \
	}; 
#endif
``

efra-mx-aqua replied:
Ignore the comment above

efra-mx-aqua replied:
This is something that would work fine:
```
#define DT_INST_ADS1X1X(inst, t) DT_INST(inst, ti_ads##t)

#define ADS1X1X_RDY_PROPS(n)                                                                       \
	.alert_rdy = GPIO_DT_SPEC_INST_GET_OR(n, alert_rdy_gpios, {0}),                              \

#define ADS1X1X_RDY(t, n)				       \
	IF_ENABLED(DT_NODE_HAS_PROP(DT_INST_ADS1X1X(n, t), alert_rdy_gpios), \
		   (ADS1X1X_RDY_PROPS(n)))

#define ADS1X1X_INIT(t, n, odr_delay_us, res, mux, pgab)                                           \
	static const struct ads1x1x_config ads##t##_config_##n = {                                 \
		.bus = I2C_DT_SPEC_GET(DT_INST_ADS1X1X(n, t)),                                     \
		.odr_delay = odr_delay_us,                                                         \
		.resolution = res,                                                                 \
		.multiplexer = mux,                                                                \
		.pga = pgab,                                                                       \
		ADS1X1X_RDY(t, n)                                                                  \
	};                                                                                         \
	static struct ads1x1x_data ads##t##_data_##n = {                                           \
		ADC_CONTEXT_INIT_LOCK(ads##t##_data_##n, ctx),                                     \
		ADC_CONTEXT_INIT_TIMER(ads##t##_data_##n, ctx),                                    \
		ADC_CONTEXT_INIT_SYNC(ads##t##_data_##n, ctx),                                     \
	};                                                                                         \
	DEVICE_DT_DEFINE(DT_INST_ADS1X1X(n, t), ads1x1x_init, NULL, &ads##t##_data_##n,            \
			 &ads##t##_config_##n, POST_KERNEL, CONFIG_ADC_ADS1X1X_INIT_PRIORITY,      \
			 &ads1x1x_api);
```

efra-mx-aqua replied:
done

XenuIsWatching commented:
Please fix the compliance issues from the CI and it appears there's another issue with complication.... other than that it looks good

XenuIsWatching reviewed (changes requested):
Please fix the compliance issues from the CI and it appears there's another issue with compiling.... other than that it looks good

XenuIsWatching commented:
You need to run on `adc_ads1x1x.c` due to the compliance checks

https://docs.zephyrproject.org/3.3.0/contribute/guidelines.html#clang-format

XenuIsWatching commented on drivers/adc/adc_ads1x1x.c:
```diff
@@ -463,13 +532,17 @@ static void adc_context_start_sampling(struct adc_context *ctx)
 {
 	int rc;
 	struct ads1x1x_data *data = dev->data;
+	const struct ads1x1x_config *config = dev->config;
```
This would give an unsed warning if CONFIG_ADC_ADS1X1X_TRIGGER is not set

efra-mx-aqua replied:
i will fix it

efra-mx-aqua replied:
done

cfriedt commented on drivers/adc/Kconfig.ads1x1x:
```diff
@@ -18,6 +18,10 @@ config ADC_ADS1X1X
 
+config ADC_ADS1X1X_TRIGGER
+	bool
+	default y
```
This should be conditional on the property being configured via Devicetree

efra-mx-aqua replied:
I will try to make some time to add this, hopefully next week

efra-mx-aqua replied:
I have introduced a new commit with a new device tree  macro `DT_ANY_COMPAT_HAS_PROP_STATUS_OKAY` 

This has been added at the top of the file

```
#if DT_ANY_COMPAT_HAS_PROP_STATUS_OKAY(ti_ads1115, alert_rdy_gpios) || \
    DT_ANY_COMPAT_HAS_PROP_STATUS_OKAY(ti_ads1114, alert_rdy_gpios) || \
    DT_ANY_COMPAT_HAS_PROP_STATUS_OKAY(ti_ads1113, alert_rdy_gpios) || \
    DT_ANY_COMPAT_HAS_PROP_STATUS_OKAY(ti_ads1014, alert_rdy_gpios) || \
    DT_ANY_COMPAT_HAS_PROP_STATUS_OKAY(ti_ads1013, alert_rdy_gpios)

#define ADC_ADS1X1X_TRIGGER

#endif
```
CONFIG_ADC_ADS1X1X_TRIGGER has been removed and replaced with `ADC_ADS1X1X_TRIGGER`

XenuIsWatching reviewed (changes requested):
See https://github.com/zephyrproject-rtos/zephyr/pull/57954#discussion_r1212827096

efra-mx-aqua committed:
devicetree: introduce DT_ANY_COMPAT_HAS_PROP_STATUS_OKAY

Add a new macro, DT_ANY_COMPAT_HAS_PROP_STATUS_OKAY
name, evaluates to 1 if any enabled instance of `compat` has the
property or to zero if it hasn't.

This macro can be useful in drivers for a family of devices.

Signed-off-by: Efrain Calderon <efrain.calderon@aquarobur.com>

efra-mx-aqua commented on drivers/adc/adc_ads1x1x.c:
```diff
@@ -20,6 +21,16 @@
 
 LOG_MODULE_REGISTER(ADS1X1X, CONFIG_ADC_LOG_LEVEL);
 
+#if DT_ANY_COMPAT_HAS_PROP_STATUS_OKAY(ti_ads1115, alert_rdy_gpios) || \
+    DT_ANY_COMPAT_HAS_PROP_STATUS_OKAY(ti_ads1114, alert_rdy_gpios) || \
+    DT_ANY_COMPAT_HAS_PROP_STATUS_OKAY(ti_ads1113, alert_rdy_gpios) || \
+    DT_ANY_COMPAT_HAS_PROP_STATUS_OKAY(ti_ads1014, alert_rdy_gpios) || \
+    DT_ANY_COMPAT_HAS_PROP_STATUS_OKAY(ti_ads1013, alert_rdy_gpios)
```
What is the recommendation here?
I think it looks ugly if they are not aligned, hehehe

XenuIsWatching replied:
I think it looks fine

XenuIsWatching reviewed (changes requested):
The gpio prop needs to be added to https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/drivers/build_all/adc/app.overlay to ensure to that it will build within the CI tests

efra-mx-aqua commented:
> The gpio prop needs to be added to https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/drivers/build_all/adc/app.overlay to ensure to that it will build within the CI tests

I have added the the gpio prop. I dont know what I should add to make sure it works when the pin is not defined. Could it be to to have two instances of each device (with and without the RDY pin)??

efra-mx-aqua commented:
> The gpio prop needs to be added to https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/drivers/build_all/adc/app.overlay to ensure to that it will build within the CI tests

I have added the the gpio prop. I dont know what I should add to make sure it works when the pin is not defined. Could it be to to have two instances of each device (with and without the RDY pin)??

efra-mx-aqua commented:
@galak @nashif do you guys have any comments ?

XenuIsWatching commented:
@efra-mx-aqua Please rebase and fix conflicts

efra-mx-aqua commented:
@efra-mx-aqua I have updated the MR.

This new version includes a fix in `ads1x1x_setup_rdy_pin()` that was causing power drain

efra-mx-aqua commented:
@vaishnavachath could you review this MR?

vaishnavachath commented on drivers/adc/adc_ads1x1x.c:
```diff
@@ -20,6 +21,17 @@
 
+#if DT_ANY_COMPAT_HAS_PROP_STATUS_OKAY(ti_ads1115, alert_rdy_gpios) || \
+	DT_ANY_COMPAT_HAS_PROP_STATUS_OKAY(ti_ads1114, alert_rdy_gpios) || \
+	DT_ANY_COMPAT_HAS_PROP_STATUS_OKAY(ti_ads1113, alert_rdy_gpios) || \
```
ADS1113 and ADS1013 does not have  ALERT/RDY pins as per the datasheet, please drop those variants if not needed.

efra-mx-aqua replied:
oops, my bad

vaishnavachath commented on drivers/adc/adc_ads1x1x.c:
```diff
@@ -138,11 +154,41 @@ struct ads1x1x_data {
+
+	gpio_pin_configure(config->alert_rdy.port,
+			   config->alert_rdy.pin,
+			   flags);
```
`gpio_pin_configure_dt()` instead? and also the return value is being ignored.

efra-mx-aqua replied:
The pin must be GPIO_DISCONNECTED, does `gpio_pin_configure_dt()` ignores the pull up/down when GPIO_DISCONNECTED is given as the flags argument?

vaishnavachath commented on drivers/adc/adc_ads1x1x.c:
```diff
@@ -138,11 +154,41 @@ struct ads1x1x_data {
+		: GPIO_INT_DISABLE;
+
+	gpio_pin_interrupt_configure(config->alert_rdy.port, config->alert_rdy.pin,
+				     flags);
```
`gpio_pin_interrupt_configure_dt()`  instead? and also the return value is being ignored.

vaishnavachath commented on include/zephyr/devicetree.h:
```diff
@@ -4051,11 +4097,29 @@
+#define DT_COMPAT_FOREACH_STATUS_OKAY_VARGS(compat, fn, ...)     \
+	COND_CODE_1(DT_HAS_COMPAT_STATUS_OKAY(compat),           \
+		    (UTIL_CAT(DT_FOREACH_OKAY_INST_VARGS_,       \
+			      compat)(fn, compat, __VA_ARGS__)), \
```
Will recommend a review from DT maintainers for this newly added helper, I think this should have a corresponding test under `tests/lib/devicetree/api/` 
CC: @cfriedt @mbolivar-ampere @mbolivar

efra-mx-aqua replied:
the test was added

mbolivar-ampere replied:
Please move this out of the instance APIs (see the ` * @defgroup devicetree-inst Instance-based devicetree APIs` above -- this section of the API is for the DT_INST_FOO() macros)

vaishnavachath commented on drivers/adc/adc_ads1x1x.c:
```diff
@@ -556,6 +649,70 @@ static void ads1x1x_acquisition_thread(const struct device *dev)
+	if (config->alert_rdy.port) {
+		ads1x1x_setup_rdy_pin(dev, false);
+		ads1x1x_setup_rdy_interrupt(dev, false);
+	}
```
Here and in other places, please consider the scenario when the GPIO/Interrupt setup/configuration fails.

efra-mx-aqua replied:
I will change that

efra-mx-aqua replied:
this is done

vaishnavachath reviewed (dismissed):
Thanks @efra-mx-aqua , except for few comments added above the changes looks okay.

mbolivar-ampere reviewed (changes requested):
Quick -1 because @vaishnavachath is correct. All new DT APIs need test cases.

Please also revisit the docstrings and make them match the patterns for the other ones in the file; this is doing something idiosyncratic.

efra-mx-aqua commented:
@mbolivar-ampere Could you be more specific about the docstring changes?

mbolivar-ampere commented on include/zephyr/devicetree.h:
```diff
@@ -4083,11 +4129,29 @@
+ *
+ *
+ * @param fn Macro to call for each enabled node. Must accept a
+ *           devicetree compatible and instance number as its only parameters.
```
This requirement is incompatible with the subsequent parameter saying that fn takes additional arguments

efra-mx replied:
fixed

mbolivar-ampere reviewed (dismissed):
Thanks for the updates -- here is a specific comment on the docstring and a request for placement.

We need test cases for all macros in this header, so the tests for DT_ANY_COMPAT_HAS_PROP_STATUS_OKAY are good, but we also need coverage for DT_COMPAT_FOREACH_STATUS_OKAY_VARGS

efra-mx-aqua commented:
Sorry, that i have not been very active but there has not been time allocated for this MR.

I am leaving my current job, so I will take over later with my personal account in my free time.

If anybody want to complete it please go ahead

efra-mx commented:
> Thanks for the updates -- here is a specific comment on the docstring and a request for placement.
> 
> We need test cases for all macros in this header, so the tests for DT_ANY_COMPAT_HAS_PROP_STATUS_OKAY are good, but we also need coverage for DT_COMPAT_FOREACH_STATUS_OKAY_VARGS

I have added a test, i hope that is enough

XenuIsWatching commented:
Please rebase

efra-mx commented:
> Please rebase

done for now.

XenuIsWatching commented:
> > Please rebase
> 
> done for now.

I don't see it rebased... it still has a merge conflict

efra-mx commented:
@XenuIsWatching well process here has been very slow. I have rebased several times, but it has become pointless since nobody is doing the review of the other files.

This MR sill become 1-year old. How could we improve that? 

I need review from @vaishnavachath @mbolivar-ampere

efra-mx commented:
> @efra-mx Thanks for this. I was forced to write my own out of tree driver a while ago to support this feature. One thing that you might consider giving some thought to with regard to this driver is the mux. Unfortunately, this driver does not support sequencing the input mux so this driver is unusable for anyone that has more than one channel connected to the ADC's inputs.

It is not unusable, but it indeed requires a different configuration, you only  need different configuration and get one channel at the time

decsny commented:
> @XenuIsWatching well process here has been very slow. I have rebased several times, but it has become pointless since nobody is doing the review of the other files.
> 
> This MR sill become 1-year old. How could we improve that?
> 
> I need review from @vaishnavachath @mbolivar-ampere

[You can follow the escalation path if needed](https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#pr-review-escalation)

sslupsky commented:
> It is not unusable, but it indeed requires a different configuration, you only need different configuration and get one channel at the time

I am not clear by what you mean.  The driver as written only configures one input channel so this driver does not work if you ask it to configure two or more input channels.  It returns an ENOTSUP error:
https://github.com/zephyrproject-rtos/zephyr/blob/3d1a4039eabe1e72163088baefec960076aec686/drivers/adc/adc_ads1x1x.c#L300-L303

Are you suggesting that the user reconfigure the ADC manually every time?  Having to manually update the configuration for each sample means the user cannot use the adc_sequence features and must duplicate a significant amount of the sampling scaffolding that is already provided by the ADC API.

I am not sure what the Zephyr policy is regarding what elements of the ADC API are mandatory but it seems necessary to include support for adc_sequence since the adc_sequence types are always included (not conditional) in the data structure:
https://github.com/zephyrproject-rtos/zephyr/blob/1680223003355673decdf18e7fd98efd9fbd2502/drivers/adc/adc_context.h#L55C1-L74C1

sslupsky commented:
This looks incorrect:
https://github.com/efra-mx/zephyr/blame/c2eed75ed0658c445f3bac66b9159a96a3b38269/drivers/adc/adc_ads1x1x.c#L481-L482

The COMP_MODE bit sets the comparator mode, it does not disable the comparator.  The COMP_QUE bits disable the comparator.

Also, you should consider initializing the COMP_POL, COMP_LAT and COMP_QUE bits.

Should the config include support for specifying a comparator configuration in the dts for users that use the comparator function?

XenuIsWatching commented:
> This looks incorrect: https://github.com/efra-mx/zephyr/blame/c2eed75ed0658c445f3bac66b9159a96a3b38269/drivers/adc/adc_ads1x1x.c#L481-L482
> 
> The COMP_MODE bit sets the comparator mode, it does not disable the comparator. The COMP_QUE bits disable the comparator.
> 
> Also, you should consider initializing the COMP_POL, COMP_LAT and COMP_QUE bits.

Yeah... that makes sense... but that probably should be in a seperate PR

XenuIsWatching commented:
I did speak to @mbolivar-ampere over Discord, he said he doesn't see anything crazy but he'm not maintaining DT right now due to dayjob time constraints. I'll ping around on "pr-help" on Discord then

efra-mx commented:
Thank you! Hopefully it will be merged.

On Thu, Mar 21, 2024, 23:46 Ryan McClelland ***@***.***>
wrote:

> I did speak to @mbolivar-ampere <https://github.com/mbolivar-ampere> over
> Discord, he said he doesn't see anything crazy but he'm not maintaining DT
> right now due to dayjob time constraints. I'll ping around on "pr-help" on
> Discord then
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/zephyrproject-rtos/zephyr/pull/57954#issuecomment-2013978977>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB5C7W44ZHCBBR6GB7V5LWTYZNPKRAVCNFSM6AAAAAAYDYB4WWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAMJTHE3TQOJXG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

sslupsky commented:
> > This looks incorrect: https://github.com/efra-mx/zephyr/blame/c2eed75ed0658c445f3bac66b9159a96a3b38269/drivers/adc/adc_ads1x1x.c#L481-L482
> > The COMP_MODE bit sets the comparator mode, it does not disable the comparator. The COMP_QUE bits disable the comparator.
> > Also, you should consider initializing the COMP_POL, COMP_LAT and COMP_QUE bits.
> 
> Yeah... that makes sense... but that probably should be in a seperate PR

I disagree.  This PR is already a year old.  Let's get this right and fix the alert issue properly and resolve this known issue and also add the missing adc sequence that is an essential component of the ADC API.

efra-mx commented:
Would you like to take over this MR?
I have no access to hardware anymore since I have changed job.



On Fri, Mar 22, 2024, 14:28 sslupsky ***@***.***> wrote:

> This looks incorrect:
> https://github.com/efra-mx/zephyr/blame/c2eed75ed0658c445f3bac66b9159a96a3b38269/drivers/adc/adc_ads1x1x.c#L481-L482
> The COMP_MODE bit sets the comparator mode, it does not disable the
> comparator. The COMP_QUE bits disable the comparator.
> Also, you should consider initializing the COMP_POL, COMP_LAT and COMP_QUE
> bits.
>
> Yeah... that makes sense... but that probably should be in a seperate PR
>
> I disagree. This PR is already a year old. Let's get this right and fix
> the alert issue properly and resolve this known issue and also add the
> missing adc sequence that is an essential component of the ADC API.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/zephyrproject-rtos/zephyr/pull/57954#issuecomment-2015100269>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB5C7W2QYVWC5D47UHQGSPDYZQWXNAVCNFSM6AAAAAAYDYB4WWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAMJVGEYDAMRWHE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

XenuIsWatching commented:
> > > This looks incorrect: https://github.com/efra-mx/zephyr/blame/c2eed75ed0658c445f3bac66b9159a96a3b38269/drivers/adc/adc_ads1x1x.c#L481-L482
> > > The COMP_MODE bit sets the comparator mode, it does not disable the comparator. The COMP_QUE bits disable the comparator.
> > > Also, you should consider initializing the COMP_POL, COMP_LAT and COMP_QUE bits.
> > 
> > 
> > Yeah... that makes sense... but that probably should be in a seperate PR
> 
> I disagree. This PR is already a year old. Let's get this right and fix the alert issue properly and resolve this known issue and also add the missing adc sequence that is an essential component of the ADC API.

A PR should be doing one thing and that is independent of what this is doing. You are welcome to submit your own PR with that code integrated from your OOT driver, and I will take a look when you do so.

efra-mx commented:
@vaishnavachath done i have fixed that

sslupsky commented:
> A PR should be doing one thing and that is independent of what this is doing. You are welcome to submit your own PR with that code integrated from your OOT driver, and I will take a look when you do so.

I understand this PR is adding the support for the ALERT/RDY pin?  The COMP_POL, COMP_LAT and COMP_QUE configuration is essential to the proper operation of this feature.  So, at a minimum, this PR should address this particular issue.

I also suggested the mux should be addressed.  I do agree that supporting the mux is not part of the original scope of the PR.  This driver should not have been merged without this capability.  Nevertheless, this PR does move the yardstick an inch or two forward so that is progress.

Regarding your suggestion that I make a PR.  The length of time it takes to get anything merged in Zephyr is outrageous.  Frankly, I don't want to have to keep rebasing a PR for a year or two until it gets merged.  I did that already and finally closed the PR's because nobody seemed to care about the bugs in some of the drivers.  You are welcome to go pick it out of my repo and submit the PR.

decsny commented:
> > A PR should be doing one thing and that is independent of what this is doing. You are welcome to submit your own PR with that code integrated from your OOT driver, and I will take a look when you do so.
> 
> I understand this PR is adding the support for the ALERT/RDY pin?  The COMP_POL, COMP_LAT and COMP_QUE configuration is essential to the proper operation of this feature.  So, at a minimum, this PR should address this particular issue.
> 
> I also suggested the mux should be addressed.  I do agree that supporting the mux is not part of the original scope of the PR.  This driver should not have been merged without this capability.  Nevertheless, this PR does move the yardstick an inch or two forward so that is progress.
> 
> Regarding your suggestion that I make a PR.  The length of time it takes to get anything merged in Zephyr is outrageous.  Frankly, I don't want to have to keep rebasing a PR for a year or two until it gets merged.  I did that already and finally closed the PR's because nobody seemed to care about the bugs in some of the drivers.  You are welcome to go pick it out of my repo and submit the PR.

Don't know when you were having those PRs but lately the project is trying to address the review velocity, there are new rules about escalation, strict expectations of assignees, and there is a public merge queue site that the release team is constantly monitoring every day, I suggest you to try again

XenuIsWatching commented on drivers/adc/adc_ads1x1x.c:
```diff
@@ -138,11 +152,41 @@ struct ads1x1x_data {
+		? GPIO_INT_EDGE_FALLING
+		: GPIO_INT_DISABLE;
+
+	gpio_pin_interrupt_configure(config->alert_rdy.port, config->alert_rdy.pin,
```
I saw the change in a previous commit, but looks like the change where this was using `gpio_pin_interrupt_configure_dt()` and checking the return value was also lost....
The original commit was here: https://github.com/zephyrproject-rtos/zephyr/commit/d84f67dc114c6472306ecd111b2885fccb131426

Anyways I did recover your old commit with some "git fu".
You can recover with the following commands
```
git fetch https://github.com/zephyrproject-rtos/zephyr.git d84f67dc114c6472306ecd111b2885fccb131426:my_lost_branch
```
The reset your branch to that, and rebase to main

XenuIsWatching reviewed (changes requested):
see comment here: https://github.com/zephyrproject-rtos/zephyr/pull/57954#discussion_r1542165943
